rfcs/20200106-prelu/README.md

```diff
@@ -0,0 +1,76 @@
+# Introducing PReLU support in oneDNN
+
+## Introduction
+
+This paper https://arxiv.org/abs/1502.01852 introduced PReLU activation function.
+Request comes from Caffe users: "PReLU is used in some POR topologies, such as
+VNET, SRGAN. It takes most of time if the original implementation, is not good
+optimized. For example VNET's train perf with latest Intel Caffe is 2.4s, PReLU
+(has optimized by openmp) spends 429ms (not including reorder time), 17% of total
+time. If we uses oneDNN ReLU to make projection of the perf, it will come to
+~1.8s." Some other data scientist in discussion mentioned that he encouter PReLU
+issue when he tried to optimize the inference pipeline of a popular face-related
+application - PReLU tooks 47 % of time in his application.
+
+Currently in oneDNN we support ReLU activation function in eltwise operation:
+https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html. LeakyReLU can be
+achieved by specifying non zero alpha parameter. PReLU is specific variant of
+LeakyReLU, where alpha is learnable parameter. It cames with two flavours:
+- channel-wise PReLU - where number of trained alpha parameters is equal to
+channel number
+- channel-shared PReLU - where there is only one trainable alpha
+
+All popular framworks supports PReLU:
+- Caffe - https://caffe.berkeleyvision.org/tutorial/layers/prelu.html
+- Keras/tensorflow - https://www.tensorflow.org/api_docs/python/tf/keras/layers/PReLU
+- PyTorch - https://pytorch.org/docs/master/generated/torch.nn.PReLU.html
+- MxNet - https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.nn.PReLU.html
+
+Example of calculations forward/backward can be found: https://bit.ly/2TXgvzT
+
+## Proposal
+
+1. Option A - create new dedicated primitive PReLU:
+
+``` cpp
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_prelu.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training,
+    /// #dnnl_forward_inference, #dnnl_backward, and #dnnl_backward_data.
+    dnnl_prop_kind_t prop_kind;
+    /// The kind of eltwise algorithm.
+    /// #dnnl_prelu_channel_wise,
+    /// #dnnl_prelu_channel_shared.
+    dnnl_alg_kind_t alg_kind;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t data_desc;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t alpha_desc;
+    /// Source and destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_data_desc;
+    /// Alpha gradient memory descriptor.
+    dnnl_memory_desc_t diff_alpha_desc;
+
+} dnnl_prelu_desc_t;
+```
+Extend dnnl_alg_kind_t: dnnl_prelu_channel_wise, dnnl_prelu_channel_shared
```

Comment:
I think we should control the algorithm (single alpha, per-channel, or (TF-variant) the whole tensor) by reading the alpha descriptor with broadcast semantics:

``` cpp
md data = {{1, 2, 3, 4}, f32, format_tag};

// per tensor alpha (aka channel shared)
md alpha_per_tensor = {{1, 1, 1, 1}, f32, format_tag::any};

// per channel alpha (aka channel-wise)
md alpha_per_channel = {{1, 2, 1, 1}, f32, format_tag::any};

// per element alpha (whole tensor, TF/Keras-variant)
md alpha_per_element = {{1, 2, 3, 4}, f32, format_tag::any};
```

Reply:
BTW, if we go this way, please specify the `format_tag::any` behavior for alpha: I think that would be the recommended option, so that we can use the format that would repeat the tensors format. Of course for single and per-channel cases this doesn't matter match, as the format will essentially be `abcd`.

Reply:
I added paragraph and open question regarding broadcast, please review if our understanding of broadcast is the same.

Reply:
Resolved in favor of https://github.com/oneapi-src/oneDNN/pull/737#discussion_r436106278